components/admin-panel/sections/ActivityLog/CollectiveEditedDetails.js

```diff
@@ -0,0 +1,112 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { diffArrays, diffChars, diffJson } from 'diff';
+import { has, isEmpty, pickBy, startCase } from 'lodash';
+import { FormattedMessage } from 'react-intl';
+import styled from 'styled-components';
+
+const diffValues = (prevValue, newValue) => {
+  if (typeof prevValue === 'string' || typeof newValue === 'string') {
+    return { type: 'string', diff: diffChars(prevValue ?? '', newValue ?? '') };
+  } else if (Array.isArray(prevValue) || Array.isArray(newValue)) {
+    return { type: 'array', diff: diffArrays(prevValue ?? [], newValue ?? []) };
+  } else if (typeof prevValue === 'object' || typeof newValue === 'object') {
+    return { type: 'object', diff: diffJson(prevValue ?? {}, newValue ?? {}) };
+  } else {
+    return {
+      type: 'default',
+      diff: [
+        { removed: true, value: JSON.stringify(prevValue) },
+        { added: true, value: JSON.stringify(newValue) },
+      ],
+    };
+  }
+};
+
+const deepCompare = (prev, next) => {
+  const removedKeys = Object.keys(prev).filter(key => !has(next, key));
+  const addedKeys = Object.keys(next).filter(key => !has(prev, key));
+  const changedValues = pickBy(next, (value, key) => !addedKeys.includes(key) && prev[key] !== value);
+  return [
+    ...removedKeys.map(key => ({ action: 'remove', key, prevValue: prev[key] })),
+    ...addedKeys.map(key => ({ action: 'add', key, newValue: next[key] })),
+    ...Object.keys(changedValues).map(key => ({
+      action: 'update',
+      key,
+      newValue: next[key],
+      prevValue: prev[key],
+      changes: diffValues(prev[key], next[key]),
+    })),
+  ];
+};
+
+const DiffLine = styled.div`
+  display: flex;
+  margin: 8px 0;
+`;
+
+const RemovedValue = styled.span`
+  background-color: #8b0000;
+  text-decoration: line-through;
+  opacity: 0.45;
+`;
+
+const AddedValue = styled.span`
+  background: #006330;
+`;
+
+const DiffedKey = styled.span`
+  font-weight: bold;
+  color: orange;
+  min-width: 100px;
+  margin-right: 8px;
+  display: block;
+`;
+
+export const CollectiveEditedDetails = ({ activity }) => {
+  const { newData, previousData } = activity.data ?? {};
+  const fullDiff = React.useMemo(() => deepCompare(previousData, newData), [newData, previousData]);
+
+  if (!fullDiff.length || (isEmpty(newData) && isEmpty(previousData))) {
+    return (
+      <i>
+        <FormattedMessage defaultMessage="No details to show" />
+      </i>
+    );
+  }
```

Comment:
I was wondering, can't we do this same check on `ActivityListItem` component and not show the `Show Details` button (or just directly add a info message saying there's no diff to show). This way the user doesn't have to click on `Show Details` and then find out  there's no details. Rather they don't have to click at all. 🙂

Reply:
It's not supposed to happen; the way you've implemented it `prevData`/`newData` is always set. Since it's abnormal behavior, I'd rather have the frontend explicitly tell the user "There's nothing to show" than hide the button and let the user wonder why it's not there.